File: packages/node/src/resources/emails.ts
import type { MissionInbox } from '../client.js';
import type {
  BulkMessageStatus,
  MessageDetails,
  MessageDetailsInclude,
  MessageStatus,
  Paginated,
  RawMessage,
  SearchEmailHit,
  SearchEmailsParams,
  SendCapStatus,
  SendEmailParams,
  SendEmailResult,
  SendLimitStatus,
} from '../types.js';

interface SendWirePayload {
  from: string;
  to?: string[];
  cc?: string[];
  bcc?: string[];
  subject?: string;
  html_body?: string;
  plain_body?: string;
  reply_to: string;
  sender?: string;
  tag?: string;
  headers?: Record<string, string>;
  message_id?: string;
  attachments?: Array<{ name: string; content_type: string; data: string }>;
}

function toArray(value: string | string[]): string[] {
  return Array.isArray(value) ? value : [value];
}

/**
 * The `emails` resource. Access via `mi.emails`.
 */
export class Emails {
  /** @internal */
  constructor(private readonly client: MissionInbox) {}

  /**
   * Send a transactional email.
   *
   * @example
   * ```ts
   * await mi.emails.send({
   *   from: 'notifications@acme.com',
   *   to: 'user@example.com',
   *   subject: 'Welcome',
   *   html: '<p>Hi</p>',
   * });
   * ```
   */
  async send(params: SendEmailParams): Promise<SendEmailResult> {
    const payload: SendWirePayload = {
      from: params.from,
      reply_to: params.replyTo ?? params.from,
    };

    if (params.to !== undefined) payload.to = toArray(params.to);
    if (params.cc !== undefined) payload.cc = toArray(params.cc);
    if (params.bcc !== undefined) payload.bcc = toArray(params.bcc);
    if (params.subject !== undefined) payload.subject = params.subject;
    if (params.html !== undefined) payload.html_body = params.html;
    if (params.text !== undefined) payload.plain_body = params.text;
    if (params.sender !== undefined) payload.sender = params.sender;
    if (params.tag !== undefined) payload.tag = params.tag;
    if (params.headers !== undefined) payload.headers = params.headers;
    if (params.messageId !== undefined) payload.message_id = params.messageId;

    if (params.attachments && params.attachments.length > 0) {
      payload.attachments = params.attachments.map((a) => ({
        name: a.filename,
        content_type: a.contentType,
        data: a.content,
      }));
    }

    return this.client.request<SendEmailResult>({
      method: 'POST',
      path: '/api/email/send',
      body: payload,
    });
  }

  /**
   * Look up the delivery status of a single message.
   *
   * @param messageId - The id returned from {@link Emails.send}.
   *
   * @example
   * ```ts
   * const status = await mi.emails.getStatus('msg_abc123');
   * if (status.bounce) console.log('bounced:', status.status);
   * ```
   */
  async getStatus(messageId: string): Promise<MessageStatus> {
    return this.client.request<MessageStatus>({
      method: 'POST',
      path: '/api/email/status',
      body: { messageId },
    });
  }

  /**
   * Look up delivery status for many messages in one request. Entries in the
   * `statuses` array align by index with the input `messageIds`. `null` entries
   * mean the id was not found.
   */
  async getBulkStatus(messageIds: string[]): Promise<BulkMessageStatus> {
    return this.client.request<BulkMessageStatus>({
      method: 'POST',
      path: '/api/email/bulk_status',
      body: { messageIds },
    });
  }

  /**
   * Fetch full details of a message. Use `include` to opt into extra sections
   * (headers, body content, attachments, activity, spam checks) — omitting it
   * returns just the top-level properties.
   *
   * @example
   * ```ts
   * const details = await mi.emails.getDetails('msg_abc123', ['content', 'activity']);
   * console.log(details.message.content?.html_body);
   * ```
   */
  async getDetails(messageId: string, include?: MessageDetailsInclude[]): Promise<MessageDetails> {
    const body: { id: string; include?: string } = { id: messageId };
    if (include && include.length > 0) body.include = include.join(',');
    return this.client.request<MessageDetails>({
      method: 'POST',
      path: '/api/email/details',
      body,
    });
  }

  /**
   * Retrieve the raw RFC-822 source of a message. Useful for debugging DKIM
   * signatures or attachments.
   */
  async getRaw(messageId: string): Promise<RawMessage> {
    return this.client.request<RawMessage>({
      method: 'POST',
      path: '/api/email/raw',
      body: { id: messageId },
    });
  }

  /**
   * Search sent/received messages. Supports partial-match filters on `from` /
   * `to`, exact match on `sendingIdentifierId`, and free-text `keyword`.
   *
   * @example
   * ```ts
   * const { data } = await mi.emails.search({
   *   from: 'notifications@',
   *   status: 'Sent',
   *   limit: 50,
   * });
   * ```
   */
  async search(params: SearchEmailsParams = {}): Promise<Paginated<SearchEmailHit>> {
    const body: Record<string, unknown> = {};
    if (params.from !== undefined) body.from = params.from;
    if (params.sendingIdentifierId !== undefined) body.sending_identifier_id = params.sendingIdentifierId;
    if (params.to !== undefined) body.to = params.to;
    if (params.messageId !== undefined) body.message_id = params.messageId;
    if (params.status !== undefined) body.status = params.status;
    if (params.direction !== undefined) body.direction = params.direction;
    if (params.keyword !== undefined) body.keyword = params.keyword;
    if (params.page !== undefined) body.page = params.page;
    if (params.limit !== undefined) body.limit = params.limit;
    if (params.order !== undefined) body.order = params.order;

    return this.client.request<Paginated<SearchEmailHit>>({
      method: 'POST',
      path: '/api/email/search',
      body,
    });
  }

  /**
   * Report the account's current send-limit state. Only the Free plan has
   * enforced caps; paid plans return `{ limited: false }`.
   */
  async getSendLimit(): Promise<SendLimitStatus> {
    return this.client.request<SendLimitStatus>({
      method: 'GET',
      path: '/api/email/send-limit-status',
    });
  }
}

// Re-exported for legibility elsewhere in the SDK.
export type { SendCapStatus };

File: packages/python/src/missioninbox/resources/emails.py
"""The `emails` resource. Access via `mi.emails`."""

from __future__ import annotations

from typing import Any, Optional, Union

from .._http import HttpClient


def _to_list(value: Union[str, list[str]]) -> list[str]:
    return list(value) if isinstance(value, list) else [value]


class Emails:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def send(
        self,
        *,
        from_: str,
        to: Union[str, list[str], None] = None,
        cc: Union[str, list[str], None] = None,
        bcc: Union[str, list[str], None] = None,
        subject: Optional[str] = None,
        html: Optional[str] = None,
        text: Optional[str] = None,
        reply_to: Optional[str] = None,
        sender: Optional[str] = None,
        tag: Optional[str] = None,
        headers: Optional[dict[str, str]] = None,
        message_id: Optional[str] = None,
        attachments: Optional[list[dict[str, str]]] = None,
    ) -> dict[str, Any]:
        """Send a transactional email.

        At least one recipient (``to``, ``cc``, or ``bcc``) and one body
        (``html`` or ``text``) are required. ``from_`` must be a registered
        sending identifier.

        Note: ``from_`` has a trailing underscore because ``from`` is a
        reserved word in Python.

        Example:
            >>> mi.emails.send(
            ...     from_="notifications@acme.com",
            ...     to="user@example.com",
            ...     subject="Welcome",
            ...     html="<p>Hi</p>",
            ... )

        Each attachment dict should have ``filename``, ``content_type``, and
        base64-encoded ``content``.
        """
        payload: dict[str, Any] = {
            "from": from_,
            "reply_to": reply_to if reply_to is not None else from_,
        }
        if to is not None:
            payload["to"] = _to_list(to)
        if cc is not None:
            payload["cc"] = _to_list(cc)
        if bcc is not None:
            payload["bcc"] = _to_list(bcc)
        if subject is not None:
            payload["subject"] = subject
        if html is not None:
            payload["html_body"] = html
        if text is not None:
            payload["plain_body"] = text
        if sender is not None:
            payload["sender"] = sender
        if tag is not None:
            payload["tag"] = tag
        if headers is not None:
            payload["headers"] = headers
        if message_id is not None:
            payload["message_id"] = message_id
        if attachments:
            payload["attachments"] = [
                {
                    "name": a["filename"],
                    "content_type": a["content_type"],
                    "data": a["content"],
                }
                for a in attachments
            ]

        return self._http.request("POST", "/api/email/send", body=payload)

    def get_status(self, message_id: str) -> dict[str, Any]:
        """Look up the delivery status of a single message."""
        return self._http.request("POST", "/api/email/status", body={"messageId": message_id})

    def get_bulk_status(self, message_ids: list[str]) -> dict[str, Any]:
        """Look up delivery status for many messages. ``None`` entries mean the id was not found."""
        return self._http.request("POST", "/api/email/bulk_status", body={"messageIds": message_ids})

    def get_details(
        self,
        message_id: str,
        include: Optional[list[str]] = None,
    ) -> dict[str, Any]:
        """Fetch full details of a message.

        ``include`` opts into extra sections: ``properties``, ``activity``,
        ``headers``, ``spam_checks``, ``content``, ``attachments``.
        """
        body: dict[str, Any] = {"id": message_id}
        if include:
            body["include"] = ",".join(include)
        return self._http.request("POST", "/api/email/details", body=body)

    def get_raw(self, message_id: str) -> dict[str, Any]:
        """Retrieve the raw RFC-822 source of a message."""
        return self._http.request("POST", "/api/email/raw", body={"id": message_id})

    def search(
        self,
        *,
        from_: Optional[str] = None,
        sending_identifier_id: Optional[str] = None,
        to: Optional[str] = None,
        message_id: Optional[str] = None,
        status: Optional[str] = None,
        direction: Optional[str] = None,
        keyword: Optional[str] = None,
        page: Optional[int] = None,
        limit: Optional[int] = None,
        order: Optional[str] = None,
    ) -> dict[str, Any]:
        """Search sent/received messages."""
        body: dict[str, Any] = {}
        if from_ is not None:
            body["from"] = from_
        if sending_identifier_id is not None:
            body["sending_identifier_id"] = sending_identifier_id
        if to is not None:
            body["to"] = to
        if message_id is not None:
            body["message_id"] = message_id
        if status is not None:
            body["status"] = status
        if direction is not None:
            body["direction"] = direction
        if keyword is not None:
            body["keyword"] = keyword
        if page is not None:
            body["page"] = page
        if limit is not None:
            body["limit"] = limit
        if order is not None:
            body["order"] = order
        return self._http.request("POST", "/api/email/search", body=body)

    def get_send_limit(self) -> dict[str, Any]:
        """Report the account's current send-limit state."""
        return self._http.request("GET", "/api/email/send-limit-status")
